File: Python3/201 - 400/400 Nth Digit.py
# 40ms, 93.20%
class Solution:
    def findNthDigit(self, n):
        base, digits = 9, 1
        while n > base * digits:
            n -= base * digits
            base *= 10
            digits += 1
        index = n % digits
        num = 10 ** (digits - 1) + n // digits
        if index == 0:
            num -= 1
            index = digits
        for _ in range(index, digits):
            num //= 10
        return num % 10
    

File: 201 - 400/400 Nth Digit.cpp
// 4ms, 58.82%
class Solution {
public:
    int findNthDigit(int n) {
        long base = 9, digits = 1;
        while (n > base * digits) {
            n -= base * digits;
            base *= 10;
            ++digits;
        }
        int index = n % digits;
        if (index == 0) index = digits;
        long num = pow(10, digits - 1);
        num += (index == digits) ? n / digits - 1 : n / digits;
        for (int i = index; i < digits; ++i)
            num /= 10;
        return num % 10;
    }
};
